### ISSUE TITLE
Unable to tell when NATs Transport is ready

### ISSUE THREAD

**Issue by dan-j** [cloudevents/sdk-go#257]
Similar to #131 but for NATS rather than HTTP.

I only started learning Go last week, so I apologise if I'm missing anything obvious.

Is there any specific reason why `StartListener` is a blocking call? It's documented within NATS `transport.go` that it's a blocking call, but the documentation on the `Transport` interface doesn't specify.

https://github.com/cloudevents/sdk-go/blob/73126cd029da7aed7b2ca33d5f4b135d8e66d051/pkg/cloudevents/transport/nats/transport.go#L116-L118

https://github.com/cloudevents/sdk-go/blob/73126cd029da7aed7b2ca33d5f4b135d8e66d051/pkg/cloudevents/transport/transport.go#L11-L22

I'd suggest changing the implementation to return after the subscription is created, something like this?

```go
t.sub, err = t.Conn.Subscribe(t.Subject, func(m *nats.Msg) {
    ...
})

go func() {
	<-ctx.Done()
	if sub != nil {
		err := sub.Unsubscribe()
		if err != nil {
			// we're in a goroutine, just log the error
		}
	}
}()

return err
```

This would be a breaking change, but I notice you haven't actually made a tagged release for this feature yet?

**Comment by n3wscott** [comment]
In general it is a blocking call so that you can do lifecycle management using the go context and not have to pass around stop channels. I think what you have suggested could be how the transport manages its self or adding some options to help with these lifecycle events.

**Comment by dan-j** [comment]
> and not have to pass around stop channels

^^ this is my point in the opposite argument, you'd have to provide a start channel which emits once the subscriber is ready.

I will look at making a separate PR as the same time as I tackle #258, as it's not much work to implement, and we can have a proper discussion once we have some actual code to review if that works for you?

**Comment by dan-j** [comment]
Just to clarify as well. I have only just started learning Go, so I'm open to being told my way isn't "the Go way", just so long as some alternative solution to the issue is agreed ☺️

**Comment by dan-j** [comment]
Just about to pick this back up and notice that v1.0.0 has been released which now makes the proposed solution a breaking change.

What's the thoughts on this? An alternative solution is to provide an additional transport `Option` to invoke a callback when the transport has established the subscription, but I'd prefer the initial solution.

**Comment by slinkydeveloper** [comment]
Is this one still valid with the latest refactoring of nats from transport to the new protocol?

**Comment by dan-j** [comment]
Yeah this isn't relevant anymore, this was back when receivers/handlers would be invoked by the subscribers, rather than consumers blocking on `receiver.Receive()`.
